spark/src/main/java/org/apache/iceberg/spark/actions/BaseRemoveFilesSparkAction.java

```diff
@@ -0,0 +1,199 @@
+  private static final String MANIFEST = "Manifest";
+  private static final String MANIFEST_LIST = "Manifest List";
+  private static final String OTHERS = "Others";
+
+  private static final String STREAM_RESULTS = "stream-results";
+
+  // Creates an executor service that runs each task in the thread that invokes execute/submit.
+  private static final ExecutorService DEFAULT_DELETE_EXECUTOR_SERVICE = null;
+
+  private final String metadataLocation;
+  private final Consumer<String> defaultDelete = new Consumer<String>() {
+    @Override
+    public void accept(String file) {
+      io.deleteFile(file);
+    }
+  };
+
+  private Consumer<String> removeFunc = defaultDelete;
+  private ExecutorService removeExecutorService = DEFAULT_DELETE_EXECUTOR_SERVICE;
+  private FileIO io = new HadoopFileIO();
+
+  public BaseRemoveFilesSparkAction(SparkSession spark, String metadataLocation) {
+    super(spark);
+    this.metadataLocation = metadataLocation;
+  }
+
+  @Override
+  protected RemoveFiles self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    if (io == null) {
+      throw new RuntimeException("IO needs to be set for removing the files");
+    }
+    String msg = String.format("Removing files reachable from %s", metadataLocation);
+    JobGroupInfo info = newJobGroupInfo("REMOVE-FILES", msg);
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
+    TableMetadata metadata = TableMetadataParser.read(io, metadataLocation);
+    Dataset<Row> validFileDF = buildValidFileDF(metadata).distinct();
+    if (streamResults) {
+      return deleteFiles(validFileDF.toLocalIterator());
+    } else {
+      return deleteFiles(validFileDF.collectAsList().iterator());
+    }
+  }
+
+  private Dataset<Row> projectFilePathWithType(Dataset<Row> ds, String type) {
+    return ds.select(functions.col("file_path"), functions.lit(type).as("file_type"));
+  }
+
+  private Dataset<Row> buildValidFileDF(TableMetadata metadata) {
+    Table staticTable = newStaticTable(metadata, io);
+    return projectFilePathWithType(buildValidDataFileDF(staticTable), DATA_FILE)
+        .union(projectFilePathWithType(buildManifestFileDF(staticTable), MANIFEST))
+        .union(projectFilePathWithType(buildManifestListDF(staticTable), MANIFEST_LIST))
+        .union(projectFilePathWithType(buildOtherMetadataFileDF(staticTable), OTHERS));
+  }
+
+  @Override
+  protected Dataset<Row> buildOtherMetadataFileDF(Table table) {
+    List<String> otherMetadataFiles = Lists.newArrayList();
+    otherMetadataFiles.addAll(ReachableFileUtil.metadataFileLocations(table, true));
+    // otherMetadataFiles.add(ReachableFileUtil.versionHintLocation(table));
+    return spark().createDataset(otherMetadataFiles, Encoders.STRING()).toDF("file_path");
+  }
+
+
+  @Override
+  public RemoveFiles io(FileIO fileIO) {
```

Comment:
It will be a bit more consistent with other actions if we `io`, `deleteWith`, `executeDeleteWith` after `self` at the top of the class. Like we have in `BaseExpireSnapshotsSparkAction`, for example.